bitbake-setup: support exporting shell environment variables via init-build-env

js731ca opened the pull request:
Rebased version of #7 

This patch adds support for exporting shell environment variables from the `init-build-env` wrapper script generated by `bitbake-setup`, based on per-configuration JSON settings.

This enables CI workflows to inject environment-specific data - such as build number, host, build type, or credentials required to fetch from certain SRC_URIs - which cannot be captured via configuration fragments alone. These variables are now handled early in the setup process and exported directly into the build environment.

**Especially tracking BB_ENV_PASSTHROUGH(_ADDITIONS) this way is interesting, since this is (currently) not possible with bitbake configuration (or fragments).**

Example:
```
  "build-environment": {
      "KEY": "value",
      "MULTILINE": [
          "line1",
          "line2"
      ]
      "ACME_DIR": "##BUILDDIR##/../layers/meta-acme",
      "BB_ENV_PASSTHROUGH_ADDITIONS": [
          "$BB_ENV_PASSTHROUGH_ADDITIONS",
          "ACME_DIR",
          "ARTIFACTORY_TOKEN",
          "ARTIFACTORY_USERNAME",
          "GITHUB_TOKEN",
          "GITHUB_PROTOCOL",
          "KEY",
          "MULTILINE"
          ]
   <snip>
```
the resulting 'init-build-env' would then be:
```
  # init-build-env wrapper created by bitbake-setup
  export KEY="value"
  export MULTILINE="line2 \
  line2"
  export ACME_DIR="/tmp/acme_master-acme-distro_acme-machine_bang/build/../layers/meta-acme"
  export BB_ENV_PASSTHROUGH_ADDITIONS="$BB_ENV_PASSTHROUGH_ADDITIONS \
  ACME_DIR \
  ARTIFACTORY_TOKEN \
  ARTIFACTORY_USERNAME \
  GITHUB_TOKEN \
  GITHUB_PROTOCOL \
  KEY \
  MULTILINE"
  . /tmp/acme_master-acme-distro_acme-machine_bang/layers/openembedded-core/oe-init-build-env /tmp/bitbake-setup/gs/acme_master-acme-distro_acme-machine_bang/build
```

js731ca committed:
bitbake-setup: support adding environment-passthroughs to the init-build-env

This patch adds support for extending the BB_ENV_PASSTHROUGH_ADDITIONS
environment variable from within the `init-build-env` wrapper script -
generated by either oe-core's oe-setup-build, or by `bitbake-setup` -
based on per-configuration JSON settings.

This enables CI workflows to inject environment-specific data - such
as build number, host, build type, or credentials required to fetch
from certain SRC_URIs - which cannot be captured via configuration
fragments alone. These variables are now handled early in the setup
process and exported directly into the build environment.

Example:

  "bb-env-passthrough-additions": [
      "ACME_DIR",
      "ARTIFACTORY_TOKEN",
      "ARTIFACTORY_USERNAME",
      "GITHUB_TOKEN",
      "GITHUB_PROTOCOL",
      "KEY"
  ]
  <snip>

the resulting 'init-build-env' would then be:
  # environment passthrough added by bitbake-setup
  export BB_ENV_PASSTHROUGH_ADDITIONS=" \
  $BB_ENV_PASSTHROUGH_ADDITIONS \
  ACME_DIR \
  ARTIFACTORY_TOKEN \
  ARTIFACTORY_USERNAME \
  GITHUB_TOKEN \
  GITHUB_PROTOCOL \
  KEY"
  # init-build-env wrapper created by bitbake-setup
  . /tmp/acme_master-acme-distro_acme-machine_bang/layers/openembedded-core/oe-init-build-env /tmp/bitbake-setup/gs/acme_master-acme-distro_acme-machine_bang/build

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

kanavin commented:
Thanks for rebasing. I have a couple things I'd like to clarify:

- the use case is injecting values that change from one CI build to another, or values that are only available in CI environment (and not in local development for example). But the implementation requires both the variable name and its value in bitbake-setup config files, which are meant to be static, and change rarely through standard human review. So how do these per-build values end up in config files in the first place? I don't understand the workflow.
- do they really need to be in a unix environment? I feel that a better path to inject them would be through build/conf/site.conf and/or build/conf/auto.conf. Is there something that prevents that?

js731ca commented:
> Thanks for rebasing. I have a couple things I'd like to clarify:
> 
>     * the use case is injecting values that change from one CI build to another, or values that are only available in CI environment (and not in local development for example). 

e.g. build-number would change from one build to another.
a login token to some "secure server" could be only available on CI (our example: a hahsicorp vault instance for keymaterial and secrets)
but a user could also have his own&separate token for local builds (our example: an artifactory instance where versioned firmware blobs live - and a user=developer has a separate accesses credentials from what the CI is using )

> But the implementation requires both the variable name and its value in bitbake-setup config files, which are meant to be static, and change rarely through standard human review. So how do these per-build values end up in config files in the first place? I don't understand the workflow.
 
i've kept the feature flexible, so that a maintainer can keep all environment related settings in one place: 
- things that are static key=value, (as single or git-blame friendly multi-line) - see GITHUB_PROTOCOL (e.g. https, ssh, ...)
- others that are just passed in from $somewhere_outside (see GITHUB_TOKEN)

>     * do they really need to be in a unix environment? I feel that a better path to inject them would be through build/conf/site.conf and/or build/conf/auto.conf. Is there something that prevents that?

did you mean injecting the VALUES?:
passing in login-tokens, passwords or other sensitive data through the environment is (AFAIK) pretty common - this way only the spawned process sees those, and they are not persisted in any (config) files somewhere. The unix-env is a "least common denominator" here, where the enclosing CI/CD does not have to know any configuration layout/scheme of the task it is running.

or the KEYs?:
then i would agree, IMHO a way to configure the BB_ENV_PASSTHROUGH through settings would be great - but what iv'e followed through the bitbake codebase this is not possible/forseen. in those places the datastore is not available yet, and/or only written to, but not read from :-(
and the "common solution" was do just pass all those settings through environment variables, and extend the bb-env-passthrough-*

kanavin commented:
> i've kept the feature flexible, so that a maintainer can keep all environment related settings in one place:
> 
>     * things that are static key=value, (as single or git-blame friendly multi-line) - see GITHUB_PROTOCOL (e.g. https, ssh, ...)
> 
>     * others that are just passed in from $somewhere_outside (see GITHUB_TOKEN)
 
But the static key=value things should simply be in a fragment. I still don't understand why they need to be in a top level bitbake-setup config and then passed in through environment (except for BB_ENV_PASSTHROUGH_ADDITIONS). Can you demonstrate how and where having them in a fragment wouldn't work? Maybe that needs to be looked at and fixed rather.
 
> > ```
> > * do they really need to be in a unix environment? I feel that a better path to inject them would be through build/conf/site.conf and/or build/conf/auto.conf. Is there something that prevents that?
> > ```
> 
> did you mean injecting the VALUES?: passing in login-tokens, passwords or other sensitive data through the environment is (AFAIK) pretty common - this way only the spawned process sees those, and they are not persisted in any (config) files somewhere. The unix-env is a "least common denominator" here, where the enclosing CI/CD does not have to know any configuration layout/scheme of the task it is running.

I understand that this is easy, but to me this doesn't look particularly secure. That spawned process is bitbake, and every sub-task it runs during a build: they would all see the sensitive data, and it could potentially be visible on the subprocesses' command lines and task logs too if handled carelessly. Having these things in files (or other secure storage) that are specifically accessible only by tools that need them would allow access control and access audit.
 
> or the KEYs?: then i would agree, IMHO a way to configure the BB_ENV_PASSTHROUGH through settings would be great - but what iv'e followed through the bitbake codebase this is not possible/forseen. in those places the datastore is not available yet, and/or only written to, but not read from :-( and the "common solution" was do just pass all those settings through environment variables, and extend the bb-env-passthrough-*

I agree that bitbake-setup should support adding to BB_ENV_PASSTHROUGH_ADDITIONS, I'm just not sure yet if it should be extended to setting environment variables in general. I'd rather have a specific configuration keyword just for that.

js731ca commented:
> > i've kept the feature flexible, so that a maintainer can keep all environment related settings in one place:
> > ```
> > * things that are static key=value, (as single or git-blame friendly multi-line) - see GITHUB_PROTOCOL (e.g. https, ssh, ...)
> > 
> > * others that are just passed in from $somewhere_outside (see GITHUB_TOKEN)
> > ```
> 
> But the static key=value things should simply be in a fragment. I still don't understand why they need to be in a top level bitbake-setup config and then passed in through environment (except for BB_ENV_PASSTHROUGH_ADDITIONS). Can you demonstrate how and where having them in a fragment wouldn't work? Maybe that needs to be looked at and fixed rather.
> 

True - config fragments should work (better) for that; my motivation was to preserve the status-quo: where most things are in the local.conf (which goes away!) and many CI related environment variables are in a (custom...) build-env-init script: BB_ENV_PASSTHROUGH_ADDTITIONS 

> > > ```
> > > * do they really need to be in a unix environment? I feel that a better path to inject them would be through build/conf/site.conf and/or build/conf/auto.conf. Is there something that prevents that?
> > > ```
> > 
> > 
> > did you mean injecting the VALUES?: passing in login-tokens, passwords or other sensitive data through the environment is (AFAIK) pretty common - this way only the spawned process sees those, and they are not persisted in any (config) files somewhere. The unix-env is a "least common denominator" here, where the enclosing CI/CD does not have to know any configuration layout/scheme of the task it is running.
> 
> I understand that this is easy, but to me this doesn't look particularly secure. That spawned process is bitbake, and every sub-task it runs during a build: they would all see the sensitive data, and it could potentially be visible on the subprocesses' command lines and task logs too if handled carelessly. Having these things in files (or other secure storage) that are specifically accessible only by tools that need them would allow access control and access audit.
>

Unfortunately true... but to fix this we would need to refactor the CI/CD codebase(s) :-S
 
> > or the KEYs?: then i would agree, IMHO a way to configure the BB_ENV_PASSTHROUGH through settings would be great - but what iv'e followed through the bitbake codebase this is not possible/forseen. in those places the datastore is not available yet, and/or only written to, but not read from :-( and the "common solution" was do just pass all those settings through environment variables, and extend the bb-env-passthrough-*
> 
> I agree that bitbake-setup should support adding to BB_ENV_PASSTHROUGH_ADDITIONS, I'm just not sure yet if it should be extended to setting environment variables in general. I'd rather have a specific configuration keyword just for that.

Okay: d6a6e65 reduces it to a new configuration element: `bb-env-passthrough-additions` which is then converted into a multi-line export statement.
(And i've already noticed a snag: this is only true for the code-path where no templates come into play -> otherwise oe-setup-build is generating it's own wrapper in build/init-build-env)

kanavin commented:
> > I agree that bitbake-setup should support adding to BB_ENV_PASSTHROUGH_ADDITIONS, I'm just not sure yet if it should be extended to setting environment variables in general. I'd rather have a specific configuration keyword just for that.
> 
> Okay: [d6a6e65](https://github.com/kanavin/bitbake/commit/d6a6e65164dc234713f66ebc8ef8ef8c3fd9424b) reduces it to a new configuration element: `bb-env-passthrough-additions` which is then converted into a multi-line export statement. (And i've already noticed a snag: this is only true for the code-path where no templates come into play -> otherwise oe-setup-build is generating it's own wrapper in build/init-build-env)

I think you can write-append to the file 'after the fact', e.g. after it was created by either bitbake-setup itself, or oe-setup-build. It's not as elegant as writing the entire content in a single transaction, but I think it should work:

```
    if template:
        bb.process.run("{} setup -c {} -b {} --no-shell".format(oesetupbuild, template, bitbake_builddir))
    else:
        initbuildenv = os.path.join(layerdir, 'init-build-env')
        if not os.path.exists(initbuildenv):
            print("Could not find oe-init-build-env in any of the layers; please use another mechanism to initialize the bitbake environment")
            return
        _make_init_build_env(bitbake_builddir, os.path.realpath(initbuildenv))

<---------------- append to file at this point

    siteconf_symlink = os.path.join(bitbake_confdir, "site.conf")
```

kanavin commented on bin/bitbake-setup:
```diff
@@ -144,8 +144,34 @@ def setup_bitbake_build(bitbake_config, layerdir, builddir):
+            f.write("# init-build-env wrapper created by bitbake-setup\n")
+            f.write(cmd + '\n')
+
+    def _prepend_passthrough_to_init_build_env(builddir):
```
Does it have to be prepend? I think appending is a bit cleaner.

js731ca replied:
you're faster than me clicking submit :-)
see below - the wrapped scripts extend&sanitize the environment variable too

kanavin commented on bin/bitbake-setup:
```diff
@@ -144,8 +144,34 @@ def setup_bitbake_build(bitbake_config, layerdir, builddir):
+            content = f.read()
+
+        if 'BB_ENV_PASSTHROUGH_ADDITIONS' in content:
+            return
```
What is the scenario when this happens, and why immediately return in this case?

js731ca replied:
... overly defensive programming; removed!
(as long as the init-build-env wrapper is always recreated we're Ok)

kanavin commented on bin/bitbake-setup:
```diff
@@ -144,8 +144,34 @@ def setup_bitbake_build(bitbake_config, layerdir, builddir):
+
+        env = bitbake_config.get("bb-env-passthrough-additions")
+        if not env:
+            return
```
This needs to be at the beginning of the function. No setting in config, nothing to do, return.

js731ca replied:
done in f9a57bb

kanavin commented on bin/bitbake-setup:
```diff
@@ -144,8 +144,34 @@ def setup_bitbake_build(bitbake_config, layerdir, builddir):
+            f.write("# environment passthrough added by bitbake-setup\n")
+            f.write(env + '\n')
+            f.write('\n')
+            f.write(content)
```
This does need an extension to existing bitbake-setup tests (which run from synthetic test data). Can you please look into it? Otherwise this code path is not exercised, and may regress without anyone noticing, until people's CI systems suddenly break on bitbake update.

js731ca replied:
ACK, see: c89808d

js731ca commented:
9abb85d follows the suggestion of extending the init-build-env script
with a "minor but": the additional export is prepended; since the wrapped `oe-init-build-env` calls down into ` scripts/oe-buildenv-internal` where the BB_ENV_PASSTHROUGH_ADDITIONS is further extended and sanitized

js731ca committed:
bitbake-setup: tests: add a template-only test configuration

bitbake-setup conf.json can be purely template driven, to setup older
yocto LTS (e.g. scarthgap) based layer-collections - where fragment
support was not yet present in oe-core+bitbake.

Add a configuration for that scenario and add it to the test routine.

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca committed:
bitbake-setup: tests: add environment-passthrough

Add a test configuration to cover the 'bb-env-passthrough-additions'
conf.json key, and add it to the test routine.

Signed-off-by: Johannes Schneider <johannes.schneider@leica-geosystems.com>

js731ca commented:
9710350 fixes a bug: `export FOO = " \` vs `export FOO=" \`

kanavin commented:
Apologies, but you need to rebase and resolve conflicts. I've made some changes request by RP on the mailing list, as that is pretty much hard high priority requirement to the thing actually landing in master, and github is again unable to make sense of it :)

kanavin commented on lib/bb/tests/setup.py:
```diff
@@ -119,6 +119,17 @@ def add_json_config_to_registry(self, name, rev, branch):
+                "bb-env-passthrough-additions": [
+                    "CORPORATE_SERVER_USERNAME",
+                    "CORPORATE_SERVER_PASSWORD",
+                    "GITUB_TOKEN"
```
Let's not set a bad example here :) BUILD_ID, BUILD_DATE, BUILD_SERVER please.

js731ca replied:
hehe - fair enough, done with the next push

kanavin commented on lib/bb/tests/setup.py:
```diff
@@ -198,6 +209,12 @@ def check_builddir_files(self, buildpath, test_file_content, json_config):
+        if 'bb-environment-passthrough' in bitbake_config.keys():
+            with open(os.path.join(bb_build_path, 'init-build-env'), 'r') as f:
+                init_build_env = f.read()
+            self.assertTrue('BB_ENV_PASSTHROUGH_ADDITIONS' in init_build_env)
```
Please also verify that the actual variables make it into the list in the file (no need to match exact value of BB_ENV_PASSTHROUGH_ADDITIONS, just that they're in the file somewhere).

js731ca replied:
ACK
